pkg/config/dumper.go

```diff
@@ -0,0 +1,90 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/bipol/scrapedumper/pkg/dumper"
+	"github.com/bipol/scrapedumper/pkg/martaapi"
+	"github.com/spf13/afero"
+	"go.uber.org/zap"
+)
+
+//DumperKind is an enum type used to specify which type of dumper is being configured
+type DumperKind string
+
+const (
+	//RoundRobinKind creates a dumper that multiplexes to several other dumpers
+	RoundRobinKind DumperKind = "ROUND_ROBIN"
+	//FileDumperKind creates a dumper that writes to a local filesystem
+	FileDumperKind DumperKind = "FILE"
+	//S3DumperKind creates a dumper that writes to an S3 bucket
+	S3DumperKind DumperKind = "S3"
+	//DynamoDBDumperKind creates a dumper that writes to a dynamodb table
+	DynamoDBDumperKind DumperKind = "DYNAMODB"
+)
+
+//DumpConfig specifies configuration for one dumper
+type DumpConfig struct {
+	Kind DumperKind `json:"kind"`
+
+	Components          []DumpConfig `json:"components"`
+	LocalOutputLocation string       `json:"local_output_location"`
+	S3BucketName        string       `json:"s3_bucket_name"`
+	DynamoTableName     string       `json:"dynamo_table_name"`
+}
+
+//BuildDumper builds the dumper described by the given config option
+func BuildDumper(log *zap.Logger, c DumpConfig) (dumper.Dumper, error) {
+	switch c.Kind {
+	case RoundRobinKind:
+		componentDumpers := make([]dumper.Dumper, len(c.Components))
+		for i := range c.Components {
+			var err error
+			componentDumpers[i], err = BuildDumper(log, c.Components[i])
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return dumper.NewRoundRobinDumpClient(log, componentDumpers...), nil
+	case FileDumperKind:
+		var localOutputLocation string
+		if c.LocalOutputLocation == "" {
+			return nil, errors.New("dumper kind FILE requested but no file output location provided: provide a local output location using the config file, a command-line argument, or an environment variable")
+		} else {
+			localOutputLocation = c.LocalOutputLocation
+		}
+
+		return dumper.NewLocalDumpHandler(localOutputLocation, log, afero.NewOsFs()), nil
+	case DynamoDBDumperKind:
+		var dynamoTableName string
+		if c.DynamoTableName == "" {
+			return nil, errors.New("dumper kind DYNAMODB requested but no dynamo table name provided: provide a dynamo table name using the config file, a command-line argument, or an environment variable")
+		} else {
+			dynamoTableName = c.DynamoTableName
+		}
+
+		dynamoClient := dynamodb.New(session.Must(session.NewSession()))
+
+		//TODO expose flexibility in the marshalFunc?
```

Comment:
q: can you add a little context to this guy, just so i can write it to a card for future work and we can remove the todo? (also fine with keeping it here tho)

Reply:
frankly, I have absolutely no idea and would love some input on that if you have any ideas! The `marshalFunc` pattern seems hard to integrate into the config struct pattern and I honestly can't think of anything non-terrible to do about that other than hard-code a specific `marshalFunc`. Maybe using a global `var` map of possible marshallers by name?